File: src/test_core/fmg/core/mosaic/MosaicControllerTest.java
package fmg.core.mosaic;

import org.junit.*;

import fmg.common.LoggerSimple;
import fmg.common.geom.Matrisize;
import fmg.common.notifier.PropertyChangeExecutor;
import fmg.core.img.IImageController;
import fmg.core.types.EGameStatus;
import fmg.core.types.EMosaic;
import fmg.core.types.EPlayInfo;
import io.reactivex.Flowable;

class MosaicTestController extends MosaicController<DummyImage, DummyImage, MosaicTestView, MosaicTestModel> {
    MosaicTestController() { super(new MosaicTestView()); }
    @Override public void close() {
        super.close();
        getView().close();
    }
}

public class MosaicControllerTest {

    /** double precision */
    static final double P = MosaicModelTest.P;

    @BeforeClass
    public static void setup() {
        LoggerSimple.put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
        LoggerSimple.put("> MosaicControllerTest::setup");

        MosaicModelTest.StaticInitializer();

        Flowable.just("UI factory inited...").subscribe(LoggerSimple::put);
    }

    @Before
    public void before() {
        LoggerSimple.put("======================================================");
    }
    @AfterClass
    public static void after() {
        LoggerSimple.put("======================================================");
        LoggerSimple.put("< MosaicControllerTest closed");
        LoggerSimple.put("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
    }

    @Test
    public void propertyChangedTest() {
        LoggerSimple.put("> MosaicControllerTest::propertyChangedTest");

        try (MosaicTestController ctrlr = new MosaicTestController()) {
            new PropertyChangeExecutor<>(ctrlr).run(100, 1000,
               () -> {
                   MosaicModelTest.changeModel(ctrlr.getModel());
               }, modifiedProperties -> {
                   Assert.assertTrue  (                    modifiedProperties.containsKey(IImageController.PROPERTY_IMAGE));
                   Assert.assertEquals(Integer.valueOf(1), modifiedProperties.get(        IImageController.PROPERTY_IMAGE).first);
               });
        }
    }

    @Test
    public void readinessAtTheStartTest() {
        LoggerSimple.put("> MosaicControllerTest::readinessAtTheStartTest");

        final int defArea = 500;
        try (MosaicTestController ctrlr = new MosaicTestController()) {
            Assert.assertEquals(defArea, ctrlr.getModel().getArea(), P);
            Assert.assertEquals(null, ctrlr.getCellDown());
            Assert.assertEquals(0, ctrlr.getCountClick());
            Assert.assertEquals(0, ctrlr.getCountFlag());
            Assert.assertEquals(10, ctrlr.getCountMinesLeft());
            Assert.assertEquals(0, ctrlr.getCountOpen());
            Assert.assertEquals(0, ctrlr.getCountUnknown());
            Assert.assertEquals(EGameStatus.eGSReady, ctrlr.getGameStatus());
            Assert.assertNotNull(ctrlr.getImage());
            Assert.assertNotNull(ctrlr.getMatrix());
            Assert.assertFalse(ctrlr.getMatrix().isEmpty());
            Assert.assertEquals(EMosaic.eMosaicSquare1, ctrlr.getMosaicType());
            Assert.assertEquals(EPlayInfo.ePlayerUnknown, ctrlr.getPlayInfo());
            Assert.assertNotNull(ctrlr.getRepositoryMines());
            Assert.assertTrue(ctrlr.getRepositoryMines().isEmpty());
            Assert.assertEquals(Math.sqrt(defArea) * 10, ctrlr.getSize().width, P);
            Assert.assertEquals(Math.sqrt(defArea) * 10, ctrlr.getSize().height, P);
            Assert.assertEquals(new Matrisize(10, 10), ctrlr.getSizeField());
        }
    }

}

File: src/test_core/fmg/core/mosaic/MosaicControllerTest.cs
using System;
using System.Threading.Tasks;
using System.Linq;
using fmg.common;
using fmg.common.geom;
using fmg.common.notifier;
using fmg.core.types;
using DummyImage = System.Object;
using MosaicTestModel = fmg.core.mosaic.MosaicDrawModel<object>;

namespace fmg.core.mosaic {

    class MosaicTestController : MosaicController<DummyImage, DummyImage, MosaicTestView, MosaicTestModel> {
        internal MosaicTestController() : base(new MosaicTestView()) { }
        protected override void Disposing() {
            base.Disposing();
            View.Dispose();
        }
    }

    public abstract class MosaicControllerTest {

        /// <summary> double precision </summary>
        private const double P = MosaicModelTest.P;

        protected abstract void AssertEqual(int    expected, int    actual);
        protected abstract void AssertEqual(double expected, double actual, double delta);
        protected abstract void AssertEqual(object expected, object actual);
        protected abstract void AssertNotNull(object anObject);
        protected abstract void AssertTrue(bool condition);
        protected abstract void AssertFalse(bool condition);

        public virtual void Setup() {
            LoggerSimple.Put(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");
            LoggerSimple.Put("> " + nameof(MosaicControllerTest) + "::" + nameof(Setup));
        }

        public virtual void Before() {
            LoggerSimple.Put("======================================================");
        }

        public virtual void After() {
            LoggerSimple.Put("======================================================");
            LoggerSimple.Put("< " + nameof(MosaicControllerTest) + " closed");
            LoggerSimple.Put("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
        }

        public virtual async Task PropertyChangedTest() {
            LoggerSimple.Put("> " + nameof(MosaicControllerTest) + "::" + nameof(PropertyChangedTest));

            await new PropertyChangeExecutor<MosaicTestController>(() => new MosaicTestController()).Run(100, 1000,
                ctrlr => {
                    MosaicModelTest.ChangeModel(ctrlr.Model);
                }, (ctrlr, modifiedProperties) => {
                    AssertTrue (   modifiedProperties.ContainsKey(nameof(ctrlr.Image)));
                    AssertEqual(1, modifiedProperties[            nameof(ctrlr.Image)]);
                });
        }

        public virtual void ReadinessAtTheStartTest() {
            LoggerSimple.Put("> " + nameof(MosaicControllerTest) + "::" + nameof(ReadinessAtTheStartTest));

            const int defArea = 500;
            using (var ctrlr = new MosaicTestController()) {
                AssertEqual(defArea, ctrlr.Model.Area, P);
                AssertEqual(null, ctrlr.CellDown);
                AssertEqual(0, ctrlr.CountClick);
                AssertEqual(0, ctrlr.CountFlag);
                AssertEqual(10, ctrlr.CountMinesLeft);
                AssertEqual(0, ctrlr.CountOpen);
                AssertEqual(0, ctrlr.CountUnknown);
                AssertEqual(EGameStatus.eGSReady, ctrlr.GameStatus);
                AssertNotNull(ctrlr.Image);
                AssertNotNull(ctrlr.Matrix);
                AssertTrue(ctrlr.Matrix.Any());
                AssertEqual(EMosaic.eMosaicSquare1, ctrlr.MosaicType);
                AssertEqual(EPlayInfo.ePlayerUnknown, ctrlr.PlayInfo);
                AssertNotNull(ctrlr.RepositoryMines);
                AssertFalse(ctrlr.RepositoryMines.Any());
                AssertEqual(Math.Sqrt(defArea) * 10, ctrlr.Size.Width, P);
                AssertEqual(Math.Sqrt(defArea) * 10, ctrlr.Size.Height, P);
                AssertEqual(new Matrisize(10, 10), ctrlr.SizeField);
            }
        }

    }

}
